server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1505,58 +1709,115 @@ private static Map<String, IndexAbstraction> filterIndicesLookupForSuffixWildcar
+                        }
+                    } else if (indexAbstraction.getType() == Type.DATA_STREAM) {
+                        DataStream dataStream = (DataStream) indexAbstraction;
+                        for (int i = 0, n = dataStream.getFailureIndices().getIndices().size(); i < n; i++) {
+                            Index index = dataStream.getFailureIndices().getIndices().get(i);
+                            IndexMetadata indexMetadata = context.state.metadata().index(index);
+                            if (indexMetadata.getState() != excludeState) {
+                                resources.add(new ResolvedExpression(index.getName(), IndexComponentSelector.DATA));
+                            }
                         }
                     }
                 }
             }
             return resources;
         }
 
-        private static List<ResolvedExpression> resolveEmptyOrTrivialWildcard(Context context) {
-            final String[] allIndices = resolveEmptyOrTrivialWildcardToAllIndices(context.getOptions(), context.getState().metadata());
-            if (context.systemIndexAccessLevel == SystemIndexAccessLevel.ALL) {
-                List<ResolvedExpression> result = new ArrayList<>(allIndices.length);
-                for (int i = 0; i < allIndices.length; i++) {
-                    result.add(new ResolvedExpression(allIndices[i]));
+        /**
+         * Adds the abstraction and selector to the results when preserving data streams and aliases at wildcard resolution. If a selector
+         * is provided, the result is only added if the selector is applicable to the abstraction provided. If
+         * {@link IndexComponentSelector#ALL_APPLICABLE} is given, the selectors are expanded only to those which are applicable to the
+         * provided abstraction.
+         * @param indexAbstraction abstraction to add
+         * @param selector The selector to add
+         * @param resources Result collector which is updated with all applicable resolved expressions for a given abstraction and selector
+         *                  pair.
+         */
+        private static void expandToApplicableSelectors(
+            IndexAbstraction indexAbstraction,
+            IndexComponentSelector selector,
+            Set<ResolvedExpression> resources
+        ) {
+            if (IndexComponentSelector.ALL_APPLICABLE.equals(selector)) {
+                resources.add(new ResolvedExpression(indexAbstraction.getName(), IndexComponentSelector.DATA));
+                if (indexAbstraction.isDataStreamRelated()) {
+                    resources.add(new ResolvedExpression(indexAbstraction.getName(), IndexComponentSelector.FAILURES));
                 }
-                return result;
+            } else if (selector == null || indexAbstraction.isDataStreamRelated() || selector.shouldIncludeFailures() == false) {
+                resources.add(new ResolvedExpression(indexAbstraction.getName(), selector));
```

Comment:
Here, it is possible to have add both: `new ResolvedExpression(indexAbstraction.getName(), DATA)` and `new ResolvedExpression(indexAbstraction.getName(), null)`. Right?

Doesn't this create a duplicate internal representation for the same resource which could cause issues in the exclusion (at least)?

Reply:
At this point in the resolution code, the only way for the selector to be null is if we are resolving expressions for an API that does not allow selectors. Otherwise, we will have assigned the default selector to be DATA. There shouldn't be a situation where we mix resolved expressions that have null and non-null selectors here.